### PR TITLE
move transactions to new file; better transaction output

### PR DESCRIPTION
**Pull request by FjorgeTheHuman**

### CODE REVIEW

**Comment by mugman174** [comment]
ok ill look

**Comment by FjorgeTheHuman** [comment]
> ok ill look

no thx lol

File: libtwirl/core/transaction.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+def transaction_commit(t): # Do the transaction itself
+	try:
+		# Start transaction
+		print("Starting transaction.")
+		t.prepare()
+		t.commit()
+	except:
+		# Handle error if package fails to install and unlock dbs
+		t.release()
+		raise Exception("Transaction failed.")
```
Comment (mugman174) [inline libtwirl/core/transaction.py:40]: You can prolly make a custom exception in libtwirl/errors

Comment (FjorgeTheHuman) [reply libtwirl/core/transaction.py:40]: I prolly can but I don't want to right now, I'm saving that for the next PR

File: libtwirl/core/install.py [hunk 2]
```diff
@@ -23,24 +24,14 @@ def pkg_install(pkgs, options = None): # Function to install packages
 		# Add the package to the list to install
 		targets.append(package)
 	# Print a list of packages to install
+	print(f"\n\033[1mPackages to install ({len(targets)}):\033[0m")
 	formatted_targets = []
 	[formatted_targets.append(f"{pkg.db.name}/{pkg.name}-{pkg.version}") for pkg in targets]
-	print('\n'.join(formatted_targets))
+	print('\n'.join(formatted_targets), "\n")
```
Comment (Sophon96) [inline libtwirl/core/install.py:30]: Note: this causes there to be one space after the line before that newline.

File: libtwirl/core/transaction.py (hunk 1 above)
Comment (caughtquick) [reply libtwirl/core/transaction.py:40]: what would that gain honestly, its cool and all but first we want to well get it to work honestly

File: libtwirl/core/install.py (hunk 2 above)
Comment (FjorgeTheHuman) [reply libtwirl/core/install.py:30]: Wdym one space? The new line after the package list is intentional

File: libtwirl/core/remove.py [hunk 3]
```diff
@@ -18,24 +19,14 @@ def pkg_remove(pkgs, options = None): # Function to uninstall packages
 		# Add the package to the list of packages to install
 		targets.append(package)
 	# Print a list of packages to install
+	print(f"\n\033[1mPackages to remove ({len(targets)}):\033[0m")
 	formatted_targets = []
 	[formatted_targets.append(f"{pkg.db.name}/{pkg.name}-{pkg.version}") for pkg in targets]
-	print('\n'.join(formatted_targets))
+	print('\n'.join(formatted_targets), "\n")
```
Comment (Sophon96) [inline libtwirl/core/remove.py:25]: Same here

**Review by mugman174 — CHANGES_REQUESTED** [review]
Just mark as resolved if you're not gonna do it

File: libtwirl/core/remove.py (hunk 3 above)
Comment (FjorgeTheHuman) [reply libtwirl/core/remove.py:25]: I still don't know what you mean

File: libtwirl/core/remove.py [hunk 4]
```diff
@@ -18,24 +19,14 @@ def pkg_remove(pkgs, options = None): # Function to uninstall packages
...
 	# Remove the packages
-	t = handle.init_transaction()
+	t = transaction.transaction_init()
 	[t.remove_pkg(pkg) for pkg in targets]
```
Comment (Sophon96) [inline libtwirl/core/remove.py:28]: IMO, use a regular `for` loop here.

File: libtwirl/core/remove.py (hunk 3 above)
Comment (mugman174) [reply libtwirl/core/remove.py:25]: Isn't it `+ "\n"` without the comma?

File: libtwirl/core/transaction.py (hunk 1 above)
Comment (mugman174) [reply libtwirl/core/transaction.py:40]: ok

File: libtwirl/core/remove.py (hunk 3 above)
Comment (FjorgeTheHuman) [reply libtwirl/core/remove.py:25]: I don't know if that would work, but I did test this PR beforehand so I know it works as I want it to

File: libtwirl/core/remove.py (hunk 4 above)
Comment (caughtquick) [reply libtwirl/core/remove.py:28]: yeah i agree, this seems like a whole lot more complexion for absolutely no reason

Comment (FjorgeTheHuman) [reply libtwirl/core/remove.py:28]: That change isn't even part of this PR, it was part of #14 iirc

File: libtwirl/core/install.py (hunk 2 above)
Comment (Sophon96) [reply libtwirl/core/install.py:30]: image

**Review by mugman174 — APPROVED** [review]
ok I'll just approve now

**Comment by caughtquick** [comment]
@Sophon96 just approve it when its done and merge it, it looks good to me with those couple of changes

File: libtwirl/core/install.py (hunk 2 above)
Comment (FjorgeTheHuman) [reply libtwirl/core/install.py:30]: oh well, you can highlight any part of the terminal you want so nobody will notice anyways

**Comment by FjorgeTheHuman** [comment]
Note: the changes to files `remove.py` and `install.py` are identical

File: libtwirl/core/transaction.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+		return default
+	else:
+		return options[option_name]
+
+def transaction_init(options = None): # Start the transaction using the options
+	# Get the handle
+	handle = database.fetch_handle()
+	# Initialise the transation and set the values
+	t = handle.init_transaction(
+		cascade=get_option_value(options, "cascade"),
```
Comment (Sophon96) [inline libtwirl/core/transaction.py:18]: IMO, we should switch to `argparse` for the frontend wrapper and call the different functions there, but this is ok for now.

**Review by Sophon96 — APPROVED** [review]

File: libtwirl/core/transaction.py (hunk 5 above)
Comment (FjorgeTheHuman) [reply libtwirl/core/transaction.py:18]: options is meant to be a dict which is passed to the action by the frontend wrapper

**Comment by caughtquick** [comment]
merging now, squashed as its the only option

File: libtwirl/core/transaction.py (hunk 5 above)
Comment (FjorgeTheHuman) [reply libtwirl/core/transaction.py:18]: each key is its own option in the dict, so I think this is fine

Comment (FjorgeTheHuman) [reply libtwirl/core/transaction.py:18]: obviously I'll make a note of it when we start writing documentation

**Pull request merged**